Firebase/Storage/FIRStorageGetDownloadURLTask.m

```diff
@@ -0,0 +1,117 @@
+ @private
+  FIRStorageVoidURLError _completion;
+}
+
+@synthesize fetcher = _fetcher;
+@synthesize fetcherCompletion = _fetcherCompletion;
+
+- (instancetype)initWithReference:(FIRStorageReference *)reference
+                   fetcherService:(GTMSessionFetcherService *)service
+                       completion:(FIRStorageVoidURLError)completion {
+  self = [super initWithReference:reference fetcherService:service];
+  if (self) {
+    _completion = [completion copy];
+  }
+  return self;
+}
+
+- (void)dealloc {
+  [_fetcher stopFetching];
+}
+
++ (NSURL *)downloadURLFromMetadataDictionary:(NSDictionary *)dictionary {
+  NSString *downloadTokens = dictionary[kFIRStorageMetadataDownloadTokens];
+
+  if (downloadTokens && downloadTokens.length > 0) {
+    NSArray<NSString *> *downloadTokenArray = [downloadTokens componentsSeparatedByString:@","];
+    NSString *bucket = dictionary[kFIRStorageMetadataBucket];
+    NSString *path = dictionary[kFIRStorageMetadataName];
+    NSString *fullPath = [NSString stringWithFormat:kFIRStorageFullPathFormat, bucket,
+                                                    [FIRStorageUtils GCSEscapedString:path]];
+
+    NSURLComponents *components = [[NSURLComponents alloc] init];
+    components.scheme = kFIRStorageScheme;
+    components.host = kFIRStorageHost;
+    components.percentEncodedPath = fullPath;
+
+    // The backend can return an arbitrary number of download tokens, but we only expose the first
+    // token via the download URL.
+    NSURLQueryItem *altItem = [[NSURLQueryItem alloc] initWithName:@"alt" value:@"media"];
+    NSURLQueryItem *tokenItem =
+        [[NSURLQueryItem alloc] initWithName:@"token" value:downloadTokenArray[0]];
+    components.queryItems = @[ altItem, tokenItem ];
+
+    return [components URL];
+  }
+
+  return nil;
+}
+
+- (void)enqueue {
+  NSMutableURLRequest *request = [self.baseRequest mutableCopy];
+  request.HTTPMethod = @"GET";
+  request.timeoutInterval = self.reference.storage.maxOperationRetryTime;
+
+  FIRStorageVoidURLError callback = _completion;
+  _completion = nil;
+
+  GTMSessionFetcher *fetcher = [self.fetcherService fetcherWithRequest:request];
+  _fetcher = fetcher;
+  fetcher.comment = @"GetDownloadURLTask";
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Warc-retain-cycles"
+  _fetcherCompletion = ^(NSData *data, NSError *error) {
+    NSURL *downloadURL;
+    if (error) {
+      if (!self.error) {
+        self.error = [FIRStorageErrors errorWithServerError:error reference:self.reference];
+      }
+    } else {
+      NSDictionary *responseDictionary = [NSDictionary frs_dictionaryFromJSONData:data];
+      if (responseDictionary != nil) {
+        downloadURL =
+            [FIRStorageGetDownloadURLTask downloadURLFromMetadataDictionary:responseDictionary];
+        if (!downloadURL) {
+          self.error =
+              [FIRStorageErrors errorWithCustomMessage:@"Failed to retrieve a download URL"];
```

Comment:
Let's go with "Failed to retrieve a download URL. Please try again."

Reply:
As discussed offline, merely added the period since retrying will likely not help.